blob/conformance.go

```diff
@@ -0,0 +1,460 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package blob provides an easy way to interact with Blob objects within
+// a bucket. It utilizes standard io packages to handle reads and writes.
+package blob
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"io/ioutil"
+	"path/filepath"
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+)
+
+// bucketMaker describes functions used to create a bucket for a test.
+// Multiple calls to bucketMaker during a test run must refer to the same
+// storage bucket.
+// Functions should return the bucket along with a "done" function to be called
+// when the test is complete.
+// If bucket creation fails, functions should t.Fatal to stop the test.
+type bucketMaker func(t *testing.T) (*Bucket, func())
+
+// RunConformanceTests runs conformance tests for provider implementations
+// of blob.
+func RunConformanceTests(t *testing.T, makeBkt bucketMaker) {
+	tests := []func(*testing.T, bucketMaker){
+		testRead,
+		testWrite,
+		testAttributes,
+		testDelete,
+	}
+	for _, testFn := range tests {
+		testFn(t, makeBkt)
+	}
+}
+
+// testRead tests the functionality of NewReader, NewRangeReader, and Reader.
+func testRead(t *testing.T, makeBkt bucketMaker) {
+	const key = "blob-for-reading"
+	content := []byte("abcdefghijklmnopqurstuvwxyz")
+	contentSize := int64(len(content))
+
+	tests := []struct {
+		name           string
+		key            string
+		offset, length int64
+		want           []byte
+		wantReadSize   int64
+		wantErr        bool
+	}{
+		{
+			name:    "read of nonexistent key fails",
+			key:     "key-does-not-exist",
+			wantErr: true,
+		},
+		{
+			name:    "negative offset fails",
+			key:     key,
+			offset:  -1,
+			wantErr: true,
+		},
+		// TODO(issue #303): Fails for GCS.
+		/*
+			{
+				name: "read metadata",
+				key:  key,
+				want: make([]byte, 0),
+			},
+		*/
+		{
+			name:         "read from positive offset to end",
+			key:          key,
+			offset:       10,
+			length:       -1,
+			want:         content[10:],
+			wantReadSize: contentSize - 10,
+		},
+		{
+			name:         "read a part in middle",
+			key:          key,
+			offset:       10,
+			length:       5,
+			want:         content[10:15],
+			wantReadSize: 5,
+		},
+		{
+			name:         "read in full",
+			key:          key,
+			offset:       0,
+			length:       -1,
+			want:         content,
+			wantReadSize: contentSize,
+		},
+	}
+
+	ctx := context.Background()
+	t.Run("TestRead", func(t *testing.T) {
```

Comment:
Push this `Run` to the caller? Decreases overall nesting.

Reply:
Done.